File: 2023/kotlin/src/Day08.kt
private enum class Instruction { LEFT, RIGHT }

private typealias Position = String
private typealias Mapping = Map<Position, MapDirection>

private data class MapDirection(val left: Position, val right: Position)

private fun parseInstructionsAndMap(input: String): Pair<List<Instruction>, Mapping> {
    val parts = input.split("\n\n")

    val instructions = parts[0].map { if (it == 'L') Instruction.LEFT else Instruction.RIGHT }

    val map = parts[1]
        .lines()
        .associate { line ->
            val (position, left, right) = Regex("([A-Z0-9]{3}).*([A-Z0-9]{3}).*([A-Z0-9]{3})").find(line)!!.destructured
            position to MapDirection(left, right)
        }

    return instructions to map
}

private fun countSteps(instructions: List<Instruction>, map: Mapping, initial: Position): Int {
    var position = initial

    instructions.cycle().forEachIndexed { steps, instruction ->
        position = if (instruction == Instruction.LEFT) map[position]!!.left else map[position]!!.right

        if (position.endsWith("Z")) {
            return steps + 1
        }
    }

    return 0
}

private fun part1(input: String) =
    parseInstructionsAndMap(input).let {
        (instructions, map) -> countSteps(instructions, map, "AAA")
    }

private fun part2(input: String): Long {
    val (instructions, map) = parseInstructionsAndMap(input)

    return map
        .keys
        .filter { it.endsWith("A") }
        .map { countSteps(instructions, map, it).toLong() }
        .reduce(::lcm)
}

fun main() {
    val input = readInput("Day08")
    part1(input).println()
    part2(input).println()
}

File: 2023/swift/AdventOfCode2023/Sources/Day08.swift
import Algorithms

struct Day08: AdventDay {
  var input: String

  typealias Position = String
  typealias Mapping = [Position: Direction]

  enum Instruction {
    case LEFT, RIGHT
  }

  struct Direction {
    let left: Position, right: Position
  }

  func parseInstructionsAndMap(_ input: String) -> ([Instruction], Mapping) {
    let parts = input.split(separator: "\n\n")
    let instructions: [Instruction] = parts[0].map { $0 == "L" ? .LEFT : .RIGHT }
    let map = "\(parts[1])"
      .lines()
      .map { line in
        let (_, position, left, right) = try! /([A-Z0-9]{3}).*([A-Z0-9]{3}).*([A-Z0-9]{3})/
          .firstMatch(in: line)!.output
        return ("\(position)", Direction(left: "\(left)", right: "\(right)"))
      }

    return (instructions, Dictionary(uniqueKeysWithValues: map))
  }

  func countSteps(_ instructions: [Instruction], _ map: Mapping, initial: Position) -> Int {
    var position = initial

    for (steps, instruction) in instructions.cycled().enumerated() {
      position = instruction == .LEFT ? map[position]!.left : map[position]!.right

      if position.last! == "Z" {
        return steps + 1
      }
    }

    return 0
  }

  func part1() -> Any {
    let (instructions, map) = parseInstructionsAndMap(input)
    return countSteps(instructions, map, initial: "AAA")
  }

  func part2() -> Any {
    func gcd(_ a: Int, _ b: Int) -> Int { b == 0 ? a : gcd(b, a % b) }
    func lcm(_ a: Int, _ b: Int) -> Int { a / gcd(a, b) * b }

    let (instructions, map) = parseInstructionsAndMap(input)

    return map
      .keys
      .filter { $0.last! == "A" }
      .map { countSteps(instructions, map, initial: $0) }
      .reduce(1, lcm)
  }
}
